core/arch/arm/kernel/thread_spmc.c

```diff
@@ -1844,26 +1849,49 @@ void thread_spmc_relinquish(uint64_t cookie)
 }
 
 static int set_pages(struct ffa_address_range *regions,
-		     unsigned int num_regions, unsigned int num_pages,
-		     struct mobj_ffa *mf)
+		     unsigned int num_regions, unsigned int *idx,
+			 struct mobj_ffa *mf)
 {
 	unsigned int n = 0;
-	unsigned int idx = 0;
 
 	for (n = 0; n < num_regions; n++) {
 		unsigned int page_count = READ_ONCE(regions[n].page_count);
 		uint64_t addr = READ_ONCE(regions[n].address);
 
-		if (mobj_ffa_add_pages_at(mf, &idx, addr, page_count))
+		if (mobj_ffa_add_pages_at(mf, idx, addr, page_count))
 			return FFA_INVALID_PARAMETERS;
 	}
 
-	if (idx != num_pages)
-		return FFA_INVALID_PARAMETERS;
-
 	return 0;
 }
 
+static struct ffa_mem_transaction *spmc_frag_retrieve(uint64_t cookie,
+										uint32_t next_fragment_offset,
+										uint32_t *fragment_length)
+{
+	struct thread_smc_args args = {
+		.a0 = FFA_MEM_FRAG_RX,
+		.a3 = next_fragment_offset,	/* offset */
+		.a4 = my_endpoint_id,		/* sender_vm_id */
+	};
+
+	reg_pair_from_64(cookie, &args.a2, &args.a1);
```

Comment:
I'm sorry, this doesn't work. I've created https://github.com/OP-TEE/optee_os/pull/6154 to provide suitable helpers to initialize `.a2` and `.a1` above instead. Until that PR is merged please revert to what you did originally.

Reply:
Done